File: VCSJones.FiddlerCertGen/Interop/LocalBufferSafeHandle.cs
using System;
using Microsoft.Win32.SafeHandles;

namespace VCSJones.FiddlerCertGen.Interop
{
    internal sealed class LocalBufferSafeHandle : SafeHandleZeroOrMinusOneIsInvalid
    {
        public LocalBufferSafeHandle(bool ownsHandle) : base(ownsHandle)
        {
        }

        public LocalBufferSafeHandle() : this(true)
        {
        }

        public static LocalBufferSafeHandle Null
        {
            get
            {
                var buffer = new LocalBufferSafeHandle(true);
                buffer.SetHandle(IntPtr.Zero);
                return buffer;
            }
        }


        public static LocalBufferSafeHandle Alloc(long size)
        {
            var native = Kernel32.LocalAlloc(0u, new IntPtr(size));
            var buffer = new LocalBufferSafeHandle(true);
            buffer.SetHandle(native);
            return buffer;
        }

        protected override bool ReleaseHandle()
        {
            return IntPtr.Zero == Kernel32.LocalFree(handle);
        }
    }
}

File: VCSJones.FidderCertGenFS/LocalBufferSafeHandle.fs
namespace VCSJones.FiddlerCertGen
open Microsoft.Win32.SafeHandles
open InteropMemoryFree

[<SealedAttribute>]
type internal LocalBufferSafeHandle =
    inherit SafeHandleZeroOrMinusOneIsInvalid

    new() = {inherit SafeHandleZeroOrMinusOneIsInvalid(true)}
    new(ownsHandle) = {inherit SafeHandleZeroOrMinusOneIsInvalid(ownsHandle)}

    static member Null
        with get () =
            let result = new LocalBufferSafeHandle()
            result.SetHandle(0n)
            result

    override this.ReleaseHandle() =
        LocalFree(this.handle) = 0n


[<SealedAttribute>]
type internal GlobalBufferSafeHandle =
    inherit SafeHandleZeroOrMinusOneIsInvalid

    new() = {inherit SafeHandleZeroOrMinusOneIsInvalid(true)}
    new(ownsHandle) = {inherit SafeHandleZeroOrMinusOneIsInvalid(ownsHandle)}

    static member Null
        with get () =
            let result = new GlobalBufferSafeHandle()
            result.SetHandle(0n)
            result

    static member FromHandle(handle : nativeint) =
        let result = new GlobalBufferSafeHandle()
        result.SetHandle(handle)
        result

    override this.ReleaseHandle() =
        System.Runtime.InteropServices.Marshal.FreeHGlobal(this.handle)
        true

